### DISCUSSION TITLE
Access request body when reusing FastAPI's exception handlers

### DISCUSSION THREAD

**Discussion by stewit** [fastapi/fastapi#8173] [Questions] [answered]
**Description**

First let me say: FastAPI is exceptionally useful to me and my team. You are doing great work here! Thank you very much.

Now to my question: **How can I access the request body when writing my own exception handlers?**

I understand that I get the request object of starlette as an argument to my own exception handler. However to access its body the coroutine request.body() must be executed (or is there any other way?). When I try to call this coroutine like in this example:

```
@app.exception_handler(RequestValidationError)
async def validation_exception_handler(request, exc):
    request_body = await request.body()
    logger.info(
        "Request validation failed:\n" + str(exc) + "\nRequest:\n" + str(request_body)
    )
    return await request_validation_exception_handler(request, exc)
```
the application hangs indefinitely / the event loop somehow freezes.

I tried to achieve the same using FastAPI's middleware decorator but that leads to the same problem.

Is it possible to to access the request body in my own exception handler? If yes, what is the recommended way?

**Comment by lvalladares** [comment]
Hey! Its a problem with starlette, check this: https://github.com/encode/starlette/issues/495#issuecomment-513138055 there is a workaround

**Comment by stewit** [comment]
Thanks to the work done in https://github.com/tiangolo/fastapi/pull/589 and the new documentation at https://fastapi.tiangolo.com/tutorial/custom-request-and-route/ I close this question.

**Comment by tiangolo** [comment]
Thanks for coming back to close the issue and for adding the link to the new docs @stewit :clap: :cake:

**Comment by janoroelandt** [comment]
The previous link to the documentation returns a 404. New location is https://fastapi.tiangolo.com/advanced/custom-request-and-route/

**Comment by YuriiMotov** [comment, marked as answer]
Another (simpler) approach is to use `body` attribute of `RequestValidationError`:

```python
@app.exception_handler(RequestValidationError)
async def validation_exception_handler(request: Request, exc: RequestValidationError):
    return JSONResponse(
        status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
        content=jsonable_encoder({"detail": exc.errors(), "body": exc.body}),
    )
```

https://fastapi.tiangolo.com/tutorial/handling-errors/#use-the-requestvalidationerror-body
